File: chapter3/028.rb
n = gets.chomp.to_i
h = gets.chomp.split(" ").map(&:to_i)

dp = Array.new(n + 1) { 0 }
h.unshift(0)

(1..n).each do |i|
  if i == 1
    dp[i] = 0
  end

  if i == 2
    dp[i] = dp[i - 1] + (h[i] - h[i - 1]).abs
  end

  if i >= 3
    tmp1 = dp[i - 1] + (h[i] - h[i - 1]).abs
    tmp2 = dp[i - 2] + (h[i] - h[i - 2]).abs
    dp[i] = [tmp1, tmp2].min
  end
end

puts dp.last

File: chapter3/028.exs
defmodule Main do
  def main do
    n = Acl.Input.get_n()
    h = Acl.Input.get_n_list()

    solve(n, h)
    |> IO.puts()
  end

  defp solve(n, h) do
    h = [0 | h]
    dp = List.duplicate(0, n + 1)

    dp =
      Enum.reduce(1..n, dp, fn i, dp ->
        cond do
          i == 1 ->
            dp

          i == 2 ->
            List.replace_at(dp, i, Enum.at(dp, i - 1) + abs(Enum.at(h, i) - Enum.at(h, i - 1)))

          true ->
            tmp1 = Enum.at(dp, i - 1) + abs(Enum.at(h, i) - Enum.at(h, i - 1))
            tmp2 = Enum.at(dp, i - 2) + abs(Enum.at(h, i) - Enum.at(h, i - 2))
            List.replace_at(dp, i, min(tmp1, tmp2))
        end
      end)

    List.last(dp)
  end
end

defmodule Acl do
  defmodule Input do
    def get_n, do: IO.read(:line) |> String.trim() |> String.to_integer()

    def get_s, do: IO.read(:line) |> String.trim()

    def get_n_list,
      do: IO.read(:line) |> String.trim() |> String.split(" ") |> Enum.map(&String.to_integer/1)

    def get_s_list,
      do: IO.read(:line) |> String.trim() |> String.split(" ")

    def get_s_split, do: IO.read(:line) |> String.trim() |> String.codepoints()
  end
end

Main.main()
